Implement sys.sequences as an empty view to fix issue in DDL Export w…

tanyagupta17 opened the pull request:
### Description

Implement sys.sequences as an empty view to fix issue in DDL Export with SSMS

Tested with SSMS 19.3 version and 18.12 version, the DDL export was successful.

Signed-off-by: Tanya Gupta [tanyagp@amazon.com](mailto:tanyagp@amazon.com)

### Issues Resolved

BABEL-5099

### Test Scenarios Covered ###
* **Use case based -** YES


* **Boundary conditions -** NA


* **Arbitrary inputs -** NA


* **Negative test cases -** NA


* **Minor version upgrade tests -** NA


* **Major version upgrade tests -** NA


* **Performance tests -** NA


* **Tooling impact -** NA


* **Client tests -** NA



### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Commit:
Implement sys.sequences as an empty view
to fix ssue in DDL Export with SSMS

Adding test cases

Signed-off-by: Tanya Gupta <tanyagp@amazon.com>

shah-nirmit commented on test/JDBC/upgrade/latest/schedule:
```diff
@@ -544,3 +544,5 @@ BABEL-3401
 BABEL-3820
```
Blocking: Please Check if we need to add DDL export tests.

tanyagupta17 replied:
As per our offline discussion: 
This is an empty view, adding a test case would return only empty view and we won't be able to generate the sequences.Neither are we adding the object in. SMO_script as it won't create any difference and is redundant for now.
Once the view is actually implemented, the DDL export tests should be added appropriately.

Commit:
Merge remote-tracking branch 'upstream/BABEL_4_X_DEV' into babel_5099_sys_sequences

Commit:
Making the test compatiable with sqlserver,removing reduntant
introduction of sequence object from SMO_script

Signed-off-by: Tanya Gupta <tanyagp@amazon.com>

rohit01010 committed:
Added wrapper function for TRIM, LTRIM, RTRIM, LEFT and RIGHT (#2481)

Currently, for string functions TRIM, LTRIM, RTRIM, LEFT and RIGHT underlying Postgres function is being used which
leads to incorrect output behaviour. This commit will fix this issue by adding wrapper functions for these string
functions.

Changes included in this commit are as follows:-

1. Added wrapper functions for TRIM, LTRIM, RTRIM, LEFT and RIGHT.
2. Added hook func_select_candidate_for_special_hook for selecting best candidate among list of functions definitions
based on return type for string functions which are in special function list.
3. Updated definition of function pltsql_report_proc_not_found_error  to include input_typeids of function.
4. Added PG_CATALOG namespace before TRIM, LTRIM, RTRIM, LEFT or RIGHT function invocations in some of the
system functions as with these wrapper definitions of TRIM, LTRIM, RTRIM, LEFT and RIGHT functions in sys
namespace, any function invocation of TRIM, LTRIM, RTRIM, LEFT or RIGHT function without namespace, by default it
will call the function in sys namespace.
4. Added new test files and updated existing test files expected output as the output behaviour of these functions will
change with this fix.
5. In case of Inline Table Valued Function(ITVF), made changes such that if the return type of body is different than
return type stored for during creation of ITVF, an implicit cast will occur and return type of body will be casted to return
type stored.
6. fixed composite actions to avoid replacing partial word for test files present in schedule file of upgrades.

Engine PR: babelfish-for-postgresql/postgresql_modified_for_babelfish#342

Task: BABEL-4489
Signed-off-by: Rohit Bhagat <rohitbgt@amazon.com>

tanyagupta17 committed:
Merge branch 'BABEL_4_X_DEV' into babel_5099_sys_sequences

thephantomthief commented on contrib/babelfishpg_tsql/sql/sys_views.sql:
```diff
@@ -3378,4 +3378,25 @@ AS SELECT
+    , CAST(0 as sys.BIT) AS is_cycling
+    , CAST(0 as sys.BIT) AS is_cached
+    , CAST(0 as INT) AS cache_size
+    , CAST(0 as sys.TINYINT) AS system_type_id
```
Although T-SQL specs say this should be tinyint, we should still keep this as int. The reason being that in future when we implement this view, we would most certainly identify this as an OID and we map OIDs to int. For example, sys.types has a column system_type_id which indicates the type oid and is of type int.

tanyagupta17 replied:
Changed from sys.TINYINT to INT as our expected OID could be larger than 255

thephantomthief commented on test/JDBC/input/views/sys_sequences-vu-prepare.sql:
```diff
@@ -0,0 +1,17 @@
+BEGIN
+    RETURN (SELECT COUNT(*) FROM sys.sequences)
+END
+GO
```
Nit: Better to have new line at the end

tanyagupta17 replied:
Done

thephantomthief commented on test/JDBC/input/views/sys_sequences-vu-verify.sql:
```diff
@@ -0,0 +1,20 @@
+sys.sequences AS seq
+ORDER BY
+[Schema] ASC,[Name] ASC
+GO
```
Nit: Better to have new line at the end

tanyagupta17 replied:
Done

Commit:
Addressed comments

Signed-off-by: Tanya Gupta <tanyagp@amazon.com>

Commit:
Merge branch 'BABEL_4_X_DEV' of https://github.com/babelfish-for-postgresql/babelfish_extensions into babel_5099_sys_sequences

Commit:
Merge branch 'babel_5099_sys_sequences' of https://github.com/amazon-aurora/babelfish_extensions into babel_5099_sys_sequences

Commit:
updating the expected file for system_type_id

Signed-off-by: Tanya Gupta <tanyagp@amazon.com>

Commit:
Added an extra blank line at end of sys_views

Signed-off-by: Tanya Gupta <tanyagp@amazon.com>

Commit:
rerun tests

Commit:
rerun tests

Commit:
Merge remote-tracking branch 'upstream/BABEL_4_X_DEV' into babel_5099_sys_sequences

Signed-off-by: Tanya Gupta <tanyagp@amazon.com>

tanyagupta17 committed:
Merge branch 'BABEL_4_X_DEV' into babel_5099_sys_sequences
